File: output/xCodeEval_remove_8000/C++/C/c2ff3a055ee6f065814ec55f552d9ddd.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

#define MAX_N 100000
#define MAX_K 100

int64_t A[MAX_N][MAX_K];
int64_t A_sum[MAX_N];
int64_t knapsack[MAX_K + 1];

int64_t max(int64_t a, int64_t b) {
    return (a > b) ? a : b;
}

void add_item(int64_t knapsack[], int size, int64_t sum, int K) {
    for (int k = K - size; k >= 0; k--) {
        knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum);
    }
}

void recurse(int start, int end, int N, int K) {
    if (start >= end) {
        return;
    }

    if (end - start == 1) {
        int64_t sum = 0;
        int64_t best = knapsack[K];

        for (int prefix = 1; prefix <= K; prefix++) {
            sum += A[start][prefix - 1];
            best = max(best, sum + knapsack[K - prefix]);
        }

        knapsack[K] = best;

        return;
    }

    int mid = (start + end) / 2;
    int64_t state[MAX_K + 1];

    for (int i = 0; i <= K; i++) {
        state[i] = knapsack[i];
    }

    for (int i = start; i < mid; i++) {
        add_item(state, K, A_sum[i], K);
    }

    recurse(mid, end, N, K);
    for (int i = 0; i <= K; i++) {
        knapsack[i] = state[i];
    }

    for (int i = mid; i < end; i++) {
        add_item(knapsack, K, A_sum[i], K);
    }

    recurse(start, mid, N, K);
}

void solve() {
    int N, K;
    scanf("%d %d", &N, &K);

    for (int i = 0; i < N; i++) {
        int t;
        scanf("%d", &t);

        for (int j = 0; j < t; j++) {
            scanf("%lld", &A[i][j]);
        }

        if (t > K) {
            t = K;
        }

        for (int j = t; j < K; j++) {
            A[i][j] = 0;
        }

        A_sum[i] = 0;
        for (int j = 0; j < K; j++) {
            A_sum[i] += A[i][j];
        }
    }

    for (int i = 0; i <= K; i++) {
        knapsack[i] = 0;
    }

    recurse(0, N, N, K);

    printf("%lld\n", knapsack[K]);
}

int main() {
    solve();
    return 0;
}

File: output/GPT4o-Mini/xCodeEval_4000/C++/Python/c2ff3a055ee6f065814ec55f552d9ddd.py
import sys
from itertools import accumulate

class DSum:
    class Solver:
        def __init__(self, in_stream, out_stream):
            self._in = in_stream
            self._out = out_stream
            self.N = 0
            self.K = 0
            self.A = []
            self.A_sum = []
            self.best = 0

        def read_ints(self):
            return list(map(int, self._in.readline().split()))

        def add_item(self, knapsack, size, sum_):
            for k in range(self.K - size, -1, -1):
                knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum_)

        def recurse(self, start, end, knapsack):
            if start >= end:
                return

            if end - start == 1:
                sum_ = 0
                self.best = max(self.best, knapsack[self.K])

                for prefix in range(1, len(self.A[start]) + 1):
                    sum_ += self.A[start][prefix - 1]
                    self.best = max(self.best, sum_ + knapsack[self.K - prefix])

                return

            mid = (start + end) // 2
            state = knapsack[:]

            for i in range(start, mid):
                self.add_item(state, len(self.A[i]), self.A_sum[i])

            self.recurse(mid, end, state)
            state = knapsack[:]

            for i in range(mid, end):
                self.add_item(state, len(self.A[i]), self.A_sum[i])

            self.recurse(start, mid, state)

        def solve(self):
            self.N, self.K = self.read_ints()
            self.A = []
            self.A_sum = []

            for _ in range(self.N):
                t, *a = self.read_ints()
                self.A.append(a)

                if t > self.K:
                    a = a[:self.K]

                self.A_sum.append(sum(a))

            self.best = 0
            knapsack = [0] * (self.K + 1)
            self.recurse(0, self.N, knapsack)
            print(self.best)

    @staticmethod
    def solve(in_stream, out_stream):
        Solver(in_stream, out_stream).solve()

if __name__ == "__main__":
    DSum.solve(sys.stdin, sys.stdout)